azkaban-common/src/main/java/azkaban/executor/ExecutableFlow.java

```diff
@@ -475,4 +482,53 @@ public VersionSet getVersionSet() {
   public void setVersionSet(final VersionSet versionSet) {
     this.versionSet = versionSet;
   }
+
+  /**
+   * Getter of flattened flow properties and flow params. This API takes lazy
+   * loading approach. If the properties are not set, then they are first set.
+   * @return Returns the flattened flow props overridden by flow params.
+   */
+  public Props getFlowPropsAndParams(final ProjectLoader projectLoader) {
+    // If the object is set, just return it.
+    if (this.flattenedFlowPropsAndParams != null) {
+      return this.flattenedFlowPropsAndParams;
+    }
+
+    Props props = new Props();
+    // Fetch the flow properties
+    if (FlowLoaderUtils.isAzkabanFlowVersion20(this.azkabanFlowVersion)) {
+      props = FlowLoaderUtils.loadPropsFromYamlFile(projectLoader,
+          this, null);
+    } else {
+      final Map<String, Props> propsMap = projectLoader.
+          fetchProjectProperties(projectId, version);
+      if (null != propsMap) {
+        Props finalProps = props;
+        propsMap.values().forEach(finalProps::putAll);
+        props = finalProps;
+      }
+    }
+
+    // Filter out the properties to keep only the override ones.
+    final Map<String, String> filteredProps = props.getMapByPrefix(PARAM_OVERRIDE);
+    final Props flowProps = new Props(null, filteredProps);
+    /*filteredProps.forEach((key, value) -> {
+      final String newKey = key.substring(PARAM_OVERRIDE.length() + 1);
+      flowProps.put(newKey, value);
+    });*/
+
+    // Fetch the flow parameters
+    Map<String, String> flowParam = null;
+    if (this.executionOptions != null) {
+      flowParam = this.executionOptions.getFlowParameters();
+    }
+    // Always put flow params AFTER the flow properties as flow params always take precedence
+    Props combinedProps = flowProps;
```

Comment:
You can avoid creating combinedProps.
Directly use this.flattenedFlowPropsAndParams = new Props(flowProps, flowParam); and return this.flattenedFlowPropsAndParams.

Reply:
ack